`panic: reflect: call of reflect.Value.Interface on zero Value` when `gh pr view --json stateReason`

### Describe the bug

```
$ gh --version
gh version 2.50.0 (2024-05-29)
https://github.com/cli/cli/releases/tag/v2.50.0
$ gh pr view --json stateReason
panic: reflect: call of reflect.Value.Interface on zero Value

goroutine 1 [running]:
reflect.valueInterface({0x0?, 0x0?, 0x0?}, 0x0?)
        reflect/value.go:1501 +0xfb
reflect.Value.Interface(...)
        reflect/value.go:1496
github.com/cli/cli/v2/api.(*PullRequest).ExportData(0xc00066d188, {0xc000958990, 0x1, 0xc0008bf438?})
        github.com/cli/cli/v2/api/export_pr.go:144 +0x968
github.com/cli/cli/v2/pkg/cmdutil.(*jsonExporter).exportData(0xc0003a4480, {0x10bdd3cc0?, 0xc00066d188?, 0xc00079cfc0?})
        github.com/cli/cli/v2/pkg/cmdutil/json_flags.go:259 +0x52a
github.com/cli/cli/v2/pkg/cmdutil.(*jsonExporter).exportData(0xc0003a4480, {0x10bd23e60?, 0xc00066d188?, 0x109915c45?})
        github.com/cli/cli/v2/pkg/cmdutil/json_flags.go:239 +0x3ac
github.com/cli/cli/v2/pkg/cmdutil.(*jsonExporter).Write(0xc0003a4480, 0xc000429e40, {0x10bd23e60, 0xc00066d188})
        github.com/cli/cli/v2/pkg/cmdutil/json_flags.go:205 +0x11c
github.com/cli/cli/v2/pkg/cmd/pr/view.viewRun(0xc000644600)
        github.com/cli/cli/v2/pkg/cmd/pr/view/view.go:121 +0x3f5
github.com/cli/cli/v2/pkg/cmd/pr/view.NewCmdView.func1(0xc00069c008, {0xc00049ebc0, 0x0, 0x10a8a9242?})
        github.com/cli/cli/v2/pkg/cmd/pr/view/view.go:69 +0x349
github.com/spf13/cobra.(*Command).execute(0xc00069c008, {0xc00049eba0, 0x2, 0x2})
        github.com/spf13/cobra@v1.8.0/command.go:983 +0xaca
github.com/spf13/cobra.(*Command).ExecuteC(0xc000459808)
        github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).ExecuteContextC(...)
        github.com/spf13/cobra@v1.8.0/command.go:1048
main.mainRun()
        github.com/cli/cli/v2/cmd/gh/main.go:119 +0x53b
main.main()
        github.com/cli/cli/v2/cmd/gh/main.go:46 +0x13
```

### Steps to reproduce the behavior

1. Check out or submit a PR
2. `gh pr view --json stateReason`
3. See error

### Expected vs actual behavior

Expected: It prints out a JSON with `stateReason`.

Actual: `gh` crashes.

### Logs

(See above)

--- a/api/query_builder.go
+++ b/api/query_builder.go
@@ -249,7 +249,7 @@ func RequiredStatusCheckRollupGraphQL(prID, after string, includeEvent bool) str
 	}`), afterClause, prID, eventField)
 }
 
-var IssueFields = []string{
+var sharedIssuePRFields = []string{
 	"assignees",
 	"author",
 	"body",
@@ -268,10 +268,20 @@ var IssueFields = []string{
 	"title",
 	"updatedAt",
 	"url",
+}
+
+// Some fields are only valid in the context of issues.
+// They need to be enumerated separately in order to be filtered
+// from existing code that expects to be able to pass Issue fields
+// to PR queries, e.g. the PullRequestGraphql function.
+var issueOnlyFields = []string{
+	"isPinned",
 	"stateReason",
 }
 
-var PullRequestFields = append(IssueFields,
+var IssueFields = append(sharedIssuePRFields, issueOnlyFields...)
+
+var PullRequestFields = append(sharedIssuePRFields,
 	"additions",
 	"autoMergeRequest",
 	"baseRefName",
@@ -298,12 +308,6 @@ var PullRequestFields = append(IssueFields,
 	"reviews",
 	"statusCheckRollup",
 )
-
-// Some fields are only valid in the context of issues.
-var issueOnlyFields = []string{
-	"isPinned",
-	"stateReason",
-}
 
 // IssueGraphQL constructs a GraphQL query fragment for a set of issue fields.
 func IssueGraphQL(fields []string) string {

--- a/pkg/cmd/issue/view/view_test.go
+++ b/pkg/cmd/issue/view/view_test.go
@@ -16,10 +16,36 @@ import (
 	"github.com/cli/cli/v2/pkg/cmdutil"
 	"github.com/cli/cli/v2/pkg/httpmock"
 	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/cli/cli/v2/pkg/jsonfieldstest"
 	"github.com/cli/cli/v2/test"
 	"github.com/google/shlex"
 	"github.com/stretchr/testify/assert"
 )
+
+func TestJSONFields(t *testing.T) {
+	jsonfieldstest.ExpectCommandToSupportJSONFields(t, NewCmdView, []string{
+		"assignees",
+		"author",
+		"body",
+		"closed",
+		"comments",
+		"createdAt",
+		"closedAt",
+		"id",
+		"labels",
+		"milestone",
+		"number",
+		"projectCards",
+		"projectItems",
+		"reactionGroups",
+		"state",
+		"title",
+		"updatedAt",
+		"url",
+		"isPinned",
+		"stateReason",
+	})
+}
 
 func runCommand(rt http.RoundTripper, isTTY bool, cli string) (*test.CmdOut, error) {
 	ios, _, stdout, stderr := iostreams.Test()

--- a/pkg/cmd/pr/view/view_test.go
+++ b/pkg/cmd/pr/view/view_test.go
@@ -18,11 +18,60 @@ import (
 	"github.com/cli/cli/v2/pkg/cmdutil"
 	"github.com/cli/cli/v2/pkg/httpmock"
 	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/cli/cli/v2/pkg/jsonfieldstest"
 	"github.com/cli/cli/v2/test"
 	"github.com/google/shlex"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
+
+func TestJSONFields(t *testing.T) {
+	jsonfieldstest.ExpectCommandToSupportJSONFields(t, NewCmdView, []string{
+		"additions",
+		"assignees",
+		"author",
+		"autoMergeRequest",
+		"baseRefName",
+		"body",
+		"changedFiles",
+		"closed",
+		"closedAt",
+		"comments",
+		"commits",
+		"createdAt",
+		"deletions",
+		"files",
+		"headRefName",
+		"headRefOid",
+		"headRepository",
+		"headRepositoryOwner",
+		"id",
+		"isCrossRepository",
+		"isDraft",
+		"labels",
+		"latestReviews",
+		"maintainerCanModify",
+		"mergeCommit",
+		"mergeStateStatus",
+		"mergeable",
+		"mergedAt",
+		"mergedBy",
+		"milestone",
+		"number",
+		"potentialMergeCommit",
+		"projectCards",
+		"projectItems",
+		"reactionGroups",
+		"reviewDecision",
+		"reviewRequests",
+		"reviews",
+		"state",
+		"statusCheckRollup",
+		"title",
+		"updatedAt",
+		"url",
+	})
+}
 
 func Test_NewCmdView(t *testing.T) {
 	tests := []struct {

--- a/pkg/jsonfieldstest/jsonfieldstest.go
+++ b/pkg/jsonfieldstest/jsonfieldstest.go
@@ -1,0 +1,47 @@
+package jsonfieldstest
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/cli/cli/v2/pkg/cmdutil"
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func jsonFieldsFor(cmd *cobra.Command) []string {
+	// This annotation is added by the `cmdutil.AddJSONFlags` function.
+	//
+	// This is an extremely janky way to get access to this information but due to the fact we pass
+	// around concrete cobra.Command structs, there's no viable way to have typed access to the fields.
+	//
+	// It's also kind of fragile because it's several hops away from the code that actually validates the usage
+	// of these flags, so it's possible for things to get out of sync.
+	stringFields, ok := cmd.Annotations["help:json-fields"]
+	if !ok {
+		return nil
+	}
+
+	return strings.Split(stringFields, ",")
+}
+
+// NewCmdFunc represents the typical function signature we use for creating commands e.g. `NewCmdView`.
+//
+// It is generic over `T` as each command construction has their own Options type e.g. `ViewOptions`
+type NewCmdFunc[T any] func(f *cmdutil.Factory, runF func(*T) error) *cobra.Command
+
+// ExpectCommandToSupportJSONFields asserts that the provided command supports exactly the provided fields.
+// Ordering of the expected fields is not important.
+//
+// Make sure you are not pointing to the same slice of fields in the test and the implementation.
+// It can be a little tedious to rewrite the fields inline in the test but it's significantly more useful because:
+//   - It forces the test author to think about and convey exactly the expected fields for a command
+//   - It avoids accidentally adding fields to a command, and the test passing unintentionally
+func ExpectCommandToSupportJSONFields[T any](t *testing.T, fn NewCmdFunc[T], expectedFields []string) {
+	t.Helper()
+
+	actualFields := jsonFieldsFor(fn(&cmdutil.Factory{}, nil))
+	assert.Equal(t, len(actualFields), len(expectedFields), "expected number of fields to match")
+	require.ElementsMatch(t, expectedFields, actualFields)
+}

